components/codec/src/byte.rs

```diff
@@ -352,12 +378,239 @@ impl MemComparableCodecHelper for DescendingMemComparableCodecHelper {
+        let mut buf = vec![0; self.bytes().len()];
+        let (read, written) = MemComparableByteCodec::try_decode_first(self.bytes(), &mut buf)?;
+        self.advance(read);
+        buf.truncate(written);
+        Ok(buf)
+    }
+}
+
+impl<T: BufferReader> MemComparableByteDecoder for T {}
+
+pub trait CompactByteEncoder {
+    /// Joins bytes with its length into a byte slice.
+    /// It is more efficient in both space and time compared to `encode_bytes`.
+    /// Note that the encoded result is not memory-comparable.
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()>;
+}
+
+impl<T: NumberEncoder> CompactByteEncoder for T {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        self.write_var_i64(data.len() as i64)?;
+        self.write_all_bytes(data)
+    }
+}
+
+impl CompactByteEncoder for std::fs::File {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        use std::io::Write;
+        let mut buf = [0; super::number::MAX_VARINT64_LENGTH];
+        let written = NumberCodec::encode_var_i64(&mut buf, data.len() as i64);
+        self.write_all(&buf[..written])?;
+        self.write_all(data)?;
+        Ok(())
+    }
+}
+
+pub trait CompactByteDecoder {
+    /// Decodes bytes which are encoded by `write_compact_bytes` before.
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>>;
+
+    /// Gets the first encoded bytes' length in compactly encoded data.
+    ///
+    /// Compact-encoding includes a VarInt encoded length prefix (1 ~ 9 bytes) and N bytes payload.
+    /// This function gets the total bytes length of compact-encoded data, including the length prefix.
+    ///
+    /// Note:
+    ///     - This function won't check whether the bytes are encoded correctly.
+    ///     - There can be multiple compact-encoded data, placed one by one. This function only returns
+    ///       the length of the first one.
+    fn encoded_compact_len(&mut self) -> usize;
+}
+
+impl<T: NumberDecoder> CompactByteDecoder for T {
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>> {
+        let vn = self.read_var_i64()? as usize;
+        let data = self.bytes();
+        if data.len() >= vn {
+            let bs = data[0..vn].to_vec();
+            self.advance(vn);
+            return Ok(bs);
+        }
+        Err(Error::UnexpectedEOF)
+    }
+
+    fn encoded_compact_len(&mut self) -> usize {
+        let last_encoded = self.bytes().as_ptr() as usize;
+        let total_len = self.bytes().len();
+        let vn = match self.read_var_i64() {
+            Ok(vn) => vn as usize,
+            Err(e) => {
+                dbg!(format!("failed to decode bytes' length: {:?}", e));
+                return total_len;
+            }
+        };
+        vn + (self.bytes().as_ptr() as usize - last_encoded)
+    }
+}
+
+impl<T: Read> CompactByteDecoder for std::io::BufReader<T> {
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>> {
```

Comment:
Where will this function being used? Looks like in existing code base we even don't need this?

Reply:
https://github.com/tikv/tikv/blob/4b48f1acafe477c9dd52720a36cfea326593b577/src/raftstore/store/snap.rs#L1000
and then 
https://github.com/tikv/tikv/blob/4b48f1acafe477c9dd52720a36cfea326593b577/src/raftstore/store/snap.rs#L825